modules/message-lane/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+		/// All stored (unprocessed) inbound messages.
+		InboundMessages: map hasher(blake2_128_concat) MessageKey => Option<T::Payload>;
+		/// Map of lane id => outbound lane data.
+		OutboundLanes: map hasher(blake2_128_concat) LaneId => OutboundLaneData;
+		/// All queued outbound messages.
+		OutboundMessages: map hasher(blake2_128_concat) MessageKey => Option<T::Payload>;
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		/// Message has been accepted and is waiting to be delivered.
+		MessageAccepted(LaneId, MessageNonce),
+		/// Messages in the inclusive range have been delivered to the bridged chain.
+		MessagesDelivered(LaneId, MessageNonce, MessageNonce),
+		/// Messages in the inclusive range have been processed by the bridged chain.
+		MessagesProcessed(LaneId, MessageNonce, MessageNonce),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+
+		/// Send message over lane.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn send_message(
+			origin,
+			lane_id: LaneId,
+			payload: T::Payload,
+		) {
+			let _ = ensure_signed(origin)?;
+			let mut lane = outbound_lane::<T, I>(lane_id);
+			let nonce = lane.send_message(payload);
+			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
+
+			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+	// =========================================================================================
+	// === Exposed mutables ====================================================================
+	// =========================================================================================
+
+	/// Receive new TRUSTED lane messages.
+	///
+	/// Trusted here means that the function itself doesn't check whether message has actually
+	/// been sent through the other end of the channel. We only check that we are receiving
+	/// and processing messages in order here.
+	///
+	/// Messages vector is required to be sorted by nonce within each lane. Otherise messages
+	/// will be rejected.
+	pub fn receive_messages(messages: Vec<Message<T::Payload>>) -> MessageNonce {
+		let mut correct_messages = 0;
+		let mut processor = T::OnMessageReceived::default();
+		for message in messages {
+			let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+			if lane.receive_message(message.key.nonce, message.payload, &mut processor) {
+				correct_messages += 1;
+			}
+		}
+
+		correct_messages
+	}
+
+	/// Process stored lane messages.
+	///
+	/// Stops processing either when all messages are processed, or when processor returns
+	/// MessageResult::NotProcessed.
+	pub fn process_lane_messages(lane_id: &LaneId, processor: &mut impl OnMessageReceived<T::Payload>) {
```

Comment:
How are we expecting this to be called? IMHO it should be part of `on_initialize` to make some progress on message processing. This ensures that we are not competing with other runtime transactions, and some scheduling mechanism can ensure fairness (for instance depending on the number of queued messages you utilize from 10%-90% of the available block weight).

Note we should require the processing fee to be payed upfront and we will probably need weight assessment of the inner-call.  Related to #316, but created a new issue to track it and discuss it https://github.com/paritytech/parity-bridges-common/issues/319